prophy/data_types.py

```diff
@@ -0,0 +1,507 @@
+from .exception import ProphyError
+from .scalar import u32, u8, scalar_bricks_walk, prophy_data_object
+from .six import repr_bytes, long
+
+
+class base_array(prophy_data_object):
+    __slots__ = ['_values']
+
+    def __init__(self):
+        self._values = []
+
+    def __getitem__(self, idx):
+        return self._values[idx]
+
+    def __getslice__(self, start, stop):
+        return self._values[start:stop]
+
+    def __len__(self):
+        return len(self._values)
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __hash__(self):
+        raise TypeError('unhashable object')
+
+    def __repr__(self):
+        return repr(self._values)
+
+    def sort(self, key_function=lambda x: x):
+        self._values.sort(key=key_function)
+
+    @classmethod
+    def _bricks_walk(cls, cursor):
+
+        if cls._max_len:
+            indexes = range(cls._max_len)
+        else:
+            if not cls._UNLIMITED:
+
+                size_info = "@%s" % cls._BOUND
+                if cls._max_len:
+                    size_info += " %s" % cls._max_len
+            else:
+                size_info = "..."
+
+            indexes = [size_info]
+
+        for index in indexes:
+            for brick_type, brick_path in cls._TYPE._bricks_walk(cursor):
+                path_ = "[%s]%s" % (index, brick_path or "")
+                yield brick_type, path_
+
+
+def distance_to_next_multiply(number, alignment):
+    remainer = number % alignment
+    return (alignment - remainer) % alignment
+
+
+def field_to_string(name, type_, value):
+    single_indent_level = " " * 2
+
+    def indent(text):
+        return '\n'.join(x and single_indent_level + x or '' for x in text.split('\n'))
+
+    if issubclass(type_, base_array):
+        return "".join(field_to_string(name, type_._TYPE, elem) for elem in value)
+    elif issubclass(type_, (struct, union)):
+        return "%s {\n%s}\n" % (name, indent(str(value)))
+    elif issubclass(type_, bytes):
+        return "%s: %s\n" % (name, repr_bytes(value))
+    elif issubclass(type_, enum):
+        return "%s: %s\n" % (name, type_._int_to_name[value])
+    else:
+        return "%s: %s\n" % (name, value)
+
+
+class _composite_base(prophy_data_object):
+    __slots__ = []
+
+    def copy_from(self, other):
+        self.validate_copy_from(other)
+        if other is self:
+            return
+
+        self._fields.clear()
+        self._copy_implementation(other)
+
+    @classmethod
+    def validate_copy_from(cls, rhs):
+        if not isinstance(rhs, cls):
+            raise TypeError("Parameter to copy_from must be instance of same class.")
+
+    def _copy_implementation(self, other):
+        " To be overrided in derived class "
+
+    @classmethod
+    def get_descriptor(cls):
+        """
+            FIXME: I'm afraid it rapes YAGNI rule
```

Comment:
* 1st: In `protobuf`:
> `ListFields(self)`
> Returns a list of (FieldDescriptor, value) tuples for all fields in the message which are not empty. 
Fields are returned with its value. The "get_descriptor" returns neither values nor paddings if any is there.

* 2nd. To determine unique identifier of the class stamp I need to iterate through each `brick` of each subclass and make it flatten. Consider this:
```
class Struct(prophy.with_metaclass(prophy.struct_generator, prophy.struct)):
    _descriptor = [("x", prophy.u32),
                   ("y", prophy.u32)]
    class Host(prophy.with_metaclass(prophy.struct_generator, prophy.struct)):
        _descriptor = [("a", Struct),
                       ("b", Struct)]
```
What if `Struct.x` changes type to say `u64`?
Iterating through `Host.get_descriptor()` will not notice any difference:
`[('a', <class 'Struct'>, ('STRUCT', 4)), ('b', <class 'Struct'>, ('STRUCT', 4))]`
Yes, then I could go recursively deeper, flatten the list, turn into strings and then meld-up with `hashlib.md5` to get an `wire interface identifier`. It's possible, but not satisfying, because my point is to also dig into bytes-stream decode/encode debugging. Then getting information about paddings becomes essential. Walk across the wire and tell/report what happened.

But going back to the `Host.Struct` case. What if the `Struct` becomes optional? The `Host.get_descriptor` will return exactly the same: `[('a', <class 'Struct'>, ('STRUCT', 4)), ('b', <class 'Struct'>, ('STRUCT', 4))]`
The optional field wil be not included.

I don't know if the `get_descriptor` is used by anybody. I assume it is and I decided to make a similar mechanism aside - iterating the "bricks_walk".

Reply:
> ListFields(self)
> Returns a list of (FieldDescriptor, value) tuples for all fields in the message which are not empty.
> Fields are returned with its value. The "get_descriptor" returns neither values nor paddings if any is there.

Yes, current `get_descriptor` is quite limited and should be enhanced. In defence of protobuf - they don't have the problem of padding, so they don't have to describe it. And when it comes to optional fields... there is some API in protobuf which allows to get a list of all fields, not only filled ones, but I don't quite remember which ;) I used protobuf as an example - to explain that it exposes publicly details about its schema in Python codec, so that users can do virtually anything with protobuf messages without necessity of changing protobuf implementation (which would be a hassle).

> But going back to the Host.Struct case. What if the Struct becomes optional? The Host.get_descriptor will return exactly the same: [('a', <class 'Struct'>, ('STRUCT', 4)), ('b', <class 'Struct'>, ('STRUCT', 4))]
> The optional field wil be not included.

This is a shortcoming which can be corrected in `get_descriptor` mechanism - it can imho return all paddings and details like: whether field is optional or not. I don't see any problem in metadata that you propose in brick mechanism - the only thing that is questionable is flattening. My point is: it's easier to flatten nested structure, than to nest flattened list, so it seems to me keeping `get_descriptor` (but enhancing it by all necessary details) - and _then_ flattening it would:
- fulfil all requirements that bricking algorithm does as it is implemented now,
- allow all kinds of usages which require nested descriptors - and that is a nice thing to have.
It's a bit more legwork, but only a bit ;)

> I don't know if the get_descriptor is used by anybody. I assume it is and I decided to make a similar mechanism aside - iterating the "bricks_walk".

I think it was added by @Jiish when he was working with Nokia as a trainee - for his master's thesis - I'm not sure if it's used in production, but I'm sure that there is a need for a fully fledged `get_descriptor` mechanism which would return all greasy details about prophy schema. Otherwise people would (and still do!) depend on private prophy details, which is obviously a wrong thing to do.

All in all - I wouldn't worry about changing `get_descriptor` API, but I think that enhancing it to a point when other uses (like walking through flattened "bricks") is possible leaves us with a prophy that is more useful for further purposes, while adding a parallel mechanism similar to `get_descriptor` makes things more confusing and harder to maintain, because there'd be two competing mechanisms for doing essentially the same thing.